hir/src/module.rs

```diff
@@ -153,6 +193,77 @@ impl Module {
         !self.link.is_linked()
     }
 
+    /// Return an iterator over the data segments allocated in this module
+    ///
+    /// The iterator is double-ended, so can be used to traverse the segments in either direction.
+    ///
+    /// Data segments are ordered by the address at which are are allocated, in ascending order.
+    pub fn segments<'a, 'b: 'a>(
+        &'b self,
+    ) -> intrusive_collections::linked_list::Iter<'a, DataSegmentAdapter> {
+        self.segments.iter()
+    }
+
+    /// Declare a new [DataSegment] in this module, with the given offset, size, and data.
+    ///
+    /// Returns `Err` if the proposed segment overlaps with an existing segment.
+    ///
+    /// Data segments are ordered by the address at which they are allocated, at link-time, all
+    /// segments from all modules are linked together, and they must either be disjoint, or exactly
+    /// identical in order to overlap - it is not permitted to have partially overlapping segments
+    /// with different views of the memory represented by that segment.
+    pub fn declare_data_segment(
+        &mut self,
+        offset: usize,
+        size: usize,
+        init: ConstantData,
+        readonly: bool,
+    ) -> Result<(), DataSegmentError> {
+        let size = core::cmp::max(size, init.len());
+        // Make sure this segment does not overlap with another segment
+        let end = offset + size;
+        let mut cursor = self.segments.front_mut();
+        while let Some(segment) = cursor.get() {
+            let segment_end = segment.offset + segment.size;
+            // If this segment starts after the segment we're declaring,
+            // we do not need to continue searching for conflicts, and
+            // can go a head and perform the insert
+            if segment.offset >= end {
+                let init = self.globals.insert_constant(init);
+                let segment = Box::new(DataSegment::new(offset, size, init, readonly));
+                cursor.insert_before(UnsafeRef::from_box(segment));
+                return Ok(());
+            }
+            // If this segment starts at the same place as the one we're
+            // declaring that's a guaranteed conflict
+            if segment.offset == offset {
+                // If the two segments have the same size and offset, then
+                // if they match in all other respects, we're done. If they
+                // don't match, then we raise a mismatch error.
+                if segment.size == size {
+                    let has_same_data = self.globals.contains_constant(&init);
+                    if has_same_data && segment.readonly == readonly {
+                        return Ok(());
+                    }
+                }
```

Comment:
When can this situation occur? Does it correspond to two globals referring to the same data or something?

It seems strange that it would be necessary to handle this corner case here. I would expect the caller to keep track of whether a particular piece of data has already been assigned a segment, and only call `declare_data_segment` when a new segment is actually needed.

Reply:
I've reworked this a bit - where the code for tracking data segments lives anyway, not this specific case. Rather than having this code live as part of `Module`, I've moved it to a separate `DataSegmentTable` data structure similar to the `GlobalVariableTable`. I'll push those changes in the morning.

But to answer your question, we want to handle that data segments might be declared in multiple modules at once, by making sure those segments don't overlap, or if they do, that they perfectly overlap (i.e. they are duplicate declarations of the same segment). This makes it trivial to take all of the segments from multiple modules and insert them into the same segment-tracking data structure without having to account for duplicates separately.